lib/spack/spack/cmd/buildcache.py

```diff
@@ -381,6 +390,47 @@ def _skip_no_redistribute_for_public(specs):
     return remaining_specs
 
 
+class PackagesAreNotInstalledError(spack.error.SpackError):
+    """Raised when a list of specs is not installed but picked to be packaged."""
+
+    def __init__(self, specs: List[Spec]):
+        super().__init__(
+            "Cannot push non-installed packages",
+            ", ".join(_format_spec(s) for s in elide_list(specs, 5)),
```

Comment:
This is the only small bug I found. `elide_list` will inject the string `...` into the list, that when passed to `_format_spec` causes:
```console
$ spack -d buildcache push --fail-fast ./mirror4
==> [2024-08-12-15:13:16.809761] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2024-08-12-15:13:16.825393] Reading config from file /home/culpo/.spack/config.yaml
==> [2024-08-12-15:13:16.829261] Reading config from file /tmp/spack-9z6krkii/spack.yaml
==> [2024-08-12-15:13:16.841737] Using environment '/tmp/spack-9z6krkii'
==> [2024-08-12-15:13:16.860899] Imported buildcache from built-in commands
==> [2024-08-12-15:13:16.864174] Imported buildcache from built-in commands
==> [2024-08-12-15:13:16.864727] Using working directory /tmp as base for abspath
==> [2024-08-12-15:13:16.864840] Using working directory /tmp as base for abspath
==> [2024-08-12-15:13:16.864919] Using working directory /tmp as base for abspath
==> [2024-08-12-15:13:16.864995] Using working directory /tmp as base for abspath
==> [2024-08-12-15:13:16.865740] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/repos.yaml
==> [2024-08-12-15:13:16.867130] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
==> [2024-08-12-15:13:17.083180] Selected 42 specs to push to file:///tmp/mirror4
==> [2024-08-12-15:13:17.083628] DATABASE LOCK TIMEOUT: 60s
==> [2024-08-12-15:13:17.083682] PACKAGE LOCK TIMEOUT: No timeout
Traceback (most recent call last):
  File "/home/culpo/PycharmProjects/spack/bin/spack", line 52, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack_installable/main.py", line 42, in main
    sys.exit(spack.main.main(argv))
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1070, in main
    return _main(argv)
           ^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1023, in _main
    return finish_parse_and_run(parser, cmd_name, args, env_format_error)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 1053, in finish_parse_and_run
    return _invoke_command(command, parser, args, unknown)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/main.py", line 650, in _invoke_command
    return_val = command(parser, args)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/buildcache.py", line 1305, in buildcache
    return args.func(args)
           ^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/buildcache.py", line 492, in push_fn
    raise PackagesAreNotInstalledError(not_installed)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/buildcache.py", line 399, in __init__
    ", ".join(_format_spec(s) for s in elide_list(specs, 5)),
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/buildcache.py", line 399, in <genexpr>
    ", ".join(_format_spec(s) for s in elide_list(specs, 5)),
              ^^^^^^^^^^^^^^^
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/cmd/buildcache.py", line 342, in _format_spec
    return spec.cformat("{name}{@version}{/hash:7}")
           ^^^^^^^^^^^^
AttributeError: 'str' object has no attribute 'cformat'. Did you mean: 'format'?

```